src/Storages/ObjectStorage/StorageObjectStorageCluster.cpp

```diff
@@ -33,14 +33,15 @@ StorageObjectStorageCluster::StorageObjectStorageCluster(
     , object_storage(object_storage_)
 {
     ColumnsDescription columns{columns_};
-    resolveSchemaAndFormat(columns, configuration->format, object_storage, configuration, {}, context_);
+    std::string sample_path;
+    resolveSchemaAndFormat(columns, configuration->format, object_storage, configuration, {}, sample_path, context_);
     configuration->check(context_);
 
     StorageInMemoryMetadata metadata;
     metadata.setColumns(columns);
     metadata.setConstraints(constraints_);
 
-    setVirtuals(VirtualColumnUtils::getVirtualsForFileLikeStorage(metadata.getColumns()));
+    setVirtuals(VirtualColumnUtils::getVirtualsForFileLikeStorage(metadata.getColumns(), context_));
```

Comment:
Actually it will be good to have the same logic with virtual columns in `s3Cluster/fileCluster/...` engines.

Reply:
I think reading part will be inherited from usual engines, so we need only to propogate sample path here (and in other cluster engines).

Reply:
We cam malke `getPathSample` method public in `StorageObjectStorage` and use `object_storage->getSamplePath` here when we didn't get `sample_path` from `resolveSchemaAndFormat`.
Would be good to add a test for it. It can look likt this:
```
select ... from s3Cluster('test_cluster_one_shard_three_replicas_localhost', 'http://localhost:11111/test/...') where ...
```